<img src="images/logo.jpg" style="display: block; margin-left: auto; margin-right: auto;" alt="לוגו של מיזם לימוד הפייתון. נחש מצויר בצבעי צהוב וכחול, הנע בין האותיות של שם הקורס: לומדים פייתון. הסלוגן המופיע מעל לשם הקורס הוא מיזם חינמי ללימוד תכנות בעברית.">

# <span style="text-align: right; direction: rtl; float: right;">Generators</span>

## <span style="text-align: right; direction: rtl; float: right; clear: both;">הקדמה</span>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    הפונקציות שיצרנו עד כה נבנו כך שיחזירו ערך אחד בכל קריאה.<br>
    הערך הזה יכול היה להיות מכל טיפוס שהוא: בוליאני, מחרוזת, tuple וכדומה.<br>
    אם נרצה להחזיר כמה ערכים יחד, תמיד נוכל להחזיר אותם כרשימה או כ־tuple.<br>
    אבל מה קורה כשאנחנו רוצים להחזיר סדרת ערכים גדולה מאוד או אפילו אין־סופית?
</p>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    למשל:
</p>

<ul style="text-align: right; direction: rtl; float: right; clear: both;">
    <li>הכתובות של כל הדפים הקיימים באינטרנט.</li>
    <li>מילות כל השירים שראו אור מאז שנת 1400 לספירה.</li>
    <li>כל המספרים השלמים הגדולים מ־0.</li>
</ul>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    בכלים שיש בידינו כרגע, נמצא שיש בעיה ליצור רשימות כאלו.<br>
    עבור רשימות גדולות מאוד – לא יהיה למחשב די זיכרון ולבסוף הוא ייכשל בשמירת ערכים חדשים.<br>
    ומה בנוגע לרשימות אין־סופיות? הן... ובכן... אין־סופיות, ולכן מלכתחילה לא נוכל ליצור אותן.
</p>

## <span style="text-align: right; direction: rtl; float: right; clear: both;">הגדרה</span>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    פתרון שמעניין לחשוב עליו הוא "פונקציה עצלנית".<br>
    אם אנחנו בשום שלב לא יכולים להחזיק בזיכרון המחשב את כל האיברים (כי יש יותר מדי מהם, או כי זו סדרה אין־סופית),<br>
    אולי נוכל לשלוח תחילה את הערך הראשון – ואת הערכים שבאים אחריו נשלח רק כשיבקשו אותם מאיתנו.
</p>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    פונקציה עצלנית שכזו נקראת <dfn>generator</dfn>, ובכל פעם שנבקש ממנה, היא תחזיר לנו איבר יחיד מתוך סדרת ערכים.<br>
    תחילה – היא תחזיר רק את הערך הראשון, בלי לחשב את שאר האיברים. אחר כך, באותו אופן, רק את השני, אחר כך רק את השלישי וכן הלאה.<br>
    ההבדל העיקרי בין generator לבין פונקציה רגילה, הוא שב־generator נבחר להחזיר את הערכים אחד־אחד ולא תחת מבנה מאוגד.
</p>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    נסכם: generator היא פונקציה שיוצרת עבורנו בכל פעם ערך אחד, מחזירה אותו, ומחכה עד שנבקש את האיבר הבא.
</p>

## <span style="text-align: right; direction: rtl; float: right; clear: both;">שימוש</span>

### <span style="text-align: right; direction: rtl; float: right; clear: both;">יצירת generator בסיסי</span>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    נתחיל בהגדרת generator מטופש למדי:
</p>

In [1]:
def silly_generator():
    a = 1
    yield a
    b = a + 1
    yield b
    c = [1, 2, 3]
    yield c

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    מעניין! זה נראה ממש כמו פונקציה. נקרא למבנה הזה שיצרנו "<dfn>פונקציית ה־generator</dfn>".<br>
    אבל מהו ה־<code>yield</code> המוזר הזה שנמצא שם?
</p>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    לפני שנתהה על קנקנו, בואו ננסה לקרוא לפונקציה ונראה מה היא מחזירה:
</p>

In [2]:
print(silly_generator())

<generator object silly_generator at 0x000001A3FCBD26C8>


<p style="text-align: right; direction: rtl; float: right; clear: both;">
    אומנם למדנו שלא אומרים איכס על פונקציות, אבל מה קורה פה?
</p>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    בניגוד לפונקציות רגילות, קריאה ל־generator לא מחזירה ערך מייד.<br>
    במקום ערך היא מחזירה מעין סמן, כמו בקובץ, שאפשר לדמיין כחץ שמצביע על השורה הראשונה של הפונקציה.<br>
    נשמור את הסמן על משתנה:
</p>

In [3]:
our_generator = silly_generator()

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    בעקבות הקריאה ל־<code>silly_generator</code> נוצר לנו סמן שמצביע כרגע על השורה <code>a = 1</code>.<br>
    המינוח המקצועי לסמן הזה הוא <dfn>generator iterator</dfn>.
</p>

<img src="images/silly_generator1.png?v=1" width="300px" style="display: block; margin-left: auto; margin-right: auto;" alt="תוכן הפונקציה silly_generator, כאשר חץ מצביע לשורה הראשונה שלה – a = 1">

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    אחרי שהרצנו את השורה <code dir="ltr">our_generator = silly_generator()</code>, הסמן המדובר נשמר במשתנה בשם <var>our_generator</var>.<br>
    זה זמן מצוין לבקש מה־generator להחזיר ערך.<br>
    נעשה זאת בעזרת הפונקציה הפייתונית <code>next</code>:
</p>

In [6]:
next_value = next(our_generator)
print(next_value)

[1, 2, 3]


<p style="text-align: right; direction: rtl; float: right; clear: both;">
    כדי להבין מה התרחש נצטרך להבין שני דברים חשובים שקשורים ל־generators:<br>
</p>
<ol style="text-align: right; direction: rtl; float: right; clear: both;">
    <li>קריאה ל־<code>next</code> היא כמו לחיצה על "נגן" (Play) – היא גורמת לסמן לרוץ עד שהוא מגיע לשורה של החזרת ערך.</li>
    <li>מילת המפתח <code>yield</code> דומה למילת המפתח <code>return</code> – היא מפסיקה את ריצת הסמן, ומחזירה את הערך שמופיע אחריה.</li>
</ol>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    אז היה לנו סמן שהצביע על השורה הראשונה. לחצנו Play, והוא הריץ את הקוד עד שהוא הגיע לנקודה שבה מחזירים ערך.<br>
    ההבדל בין פונקציה לבין generator, הוא ש<mark>כשאנחנו מחזירים ערך בעזרת <code>yield</code> אנחנו "מקפיאים" את המצב שבו
    יצאנו מהפונקציה.</mark><br>
    ממש כמו ללחוץ על "Pause".<br>
    כשנקרא ל־<code>next</code> בפעם הבאה – הפונקציה תמשיך לרוץ מאותו המקום שבו השארנו את הסמן, עם אותם ערכי משתנים.<br>
    עכשיו הסמן מצביע על השורה <code>b = a + 1</code>, ומחכה שמישהו יקרא שוב ל־<code>next</code> כדי שהפונקציה תוכל להמשיך לרוץ:
</p>

In [6]:
print(next(our_generator))

2


<p style="text-align: right; direction: rtl; float: right; clear: both;">
    נסכם מה קרה עד עכשיו:
</p>

<ol style="text-align: right; direction: rtl; float: right; clear: both;">
    <li>הגדרנו פונקציה בשם <var>silly_generator</var>, שאמורה להחזיר את הערכים <samp>1</samp>, <samp>2</samp> ו־<samp dir="ltr">[1, 2, 3]</samp>. קראנו לה "<em>פונקציית הגנרטור</em>".</li>
    <li>בעזרת קריאה לפונקציית הגנרטור, יצרנו "סמן" (generator iterator) שנקרא <var>our_generator</var> ומצביע לשורה הראשונה בפונקציה.</li>
    <li>בעזרת קריאה ל־<code>next</code> על ה־generator iterator, הרצנו את הסמן עד שה־generator החזיר ערך.</li>
    <li>למדנו ש־generator־ים מחזירים ערכים בעיקר בעזרת yield – שמחזיר ערך ושומר את המצב שבו הפונקציה עצרה.</li>
    <li>קראנו שוב ל־<code>next</code> על ה־generator iterator, וראינו שהוא ממשיך מהמקום שבו ה־generator הפסיק לרוץ פעם קודמת.</li>
</ol>

<div class="align-center" style="display: flex; text-align: right; direction: rtl; clear: both;">
    <div style="display: flex; width: 10%; float: right; clear: both;">
        <img src="images/exercise.svg" style="height: 50px !important;" alt="תרגול"> 
    </div>
    <div style="width: 70%">
        <p style="text-align: right; direction: rtl; float: right; clear: both;">
            תוכלו לחזות מה יקרה אם נקרא שוב ל־<code>next(our_generator)</code>?
        </p>
    </div>
    <div style="display: flex; width: 20%; border-right: 0.1rem solid #A5A5A5; padding: 1rem 2rem;">
        <p style="text-align: center; direction: rtl; justify-content: center; align-items: center; clear: both;">
            <strong>חשוב!</strong><br>
            פתרו לפני שתמשיכו!
        </p>
    </div>
</div>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    ננסה:
</p>

In [7]:
print(next(our_generator))

[1, 2, 3]


<p style="text-align: right; direction: rtl; float: right; clear: both;">
    יופי! הכל הלך כמצופה.<br>
    אבל מה צופן לנו העתיד?<br>
    בפעם הבאה שנבקש ערך מהפונקציה, הסמן שלנו ירוץ הלאה ולא ייתקל ב־<code>yield</code>.<br>
    במקרה כזה, נקבל שגיאת <var>StopIteration</var>, שמבשרת לנו ש־<code>next</code> לא הצליח לחלץ מה־generator את הערך הבא.
</p>

In [8]:
print(next(our_generator))

StopIteration: 

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    מובן שאין סיבה להילחץ.<br>
    במקרה הזה אפילו לא מדובר במשהו רע – פשוט כילינו את כל הערכים מה־generator iterator שלנו.<br>
    פונקציית ה־generator עדיין קיימת!<br>
    אפשר ליצור עוד generator iterator אם נרצה, ולקבל את כל הערכים שנמצאים בו באותה צורה:
</p>

In [9]:
our_generator = silly_generator()
print(next(our_generator))
print(next(our_generator))
print(next(our_generator))

1
2
[1, 2, 3]


<p style="text-align: right; direction: rtl; float: right; clear: both;">
    אבל כשחושבים על זה, זה קצת מגוחך.<br>
    בכל פעם שנרצה להשיג את הערך הבא נצטרך לרשום <code>next</code>?<br>
    חייבת להיות דרך טובה יותר!
</p>

### <span style="text-align: right; direction: rtl; float: right; clear: both;">כל generator הוא גם iterable</span>

#### <span style="text-align: right; direction: rtl; float: right; clear: both;">for</span>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    אז למעשה, יש יותר מדרך טובה אחת להשיג את כל הערכים שיוצאים מ־generator מסוים.<br>
    כהקדמה, נניח פה עובדה שלא תשאיר אתכם אדישים: ה־generator iterator הוא... iterable! הפתעת השנה, אני יודע!<br>
    אמנם אי אפשר לפנות לאיברים שלו לפי מיקום, אך בהחלט אפשר לעבור עליהם בעזרת לולאת <code>for</code>, לדוגמה:
</p>

In [10]:
our_generator = silly_generator()
for item in our_generator:
    print(item)

1
2
[1, 2, 3]


<p style="text-align: right; direction: rtl; float: right; clear: both;">
    מה מתרחש כאן?<br>
    אנחנו מבקשים מלולאת ה־<code>for</code> לעבור על ה־generator iterator שלנו.<br>
    ה־<code>for</code> עושה עבורנו את העבודה אוטומטית:
</p>

<ol style="text-align: right; direction: rtl; float: right; clear: both;">
    <li>הוא מבקש את האיבר הבא מה־generator iterator באמצעות <code>next</code>.</li>
    <li>הוא מכניס את האיבר שהוא קיבל מה־generator ל־<var>item</var>.</li>
    <li>הוא מבצע את גוף הלולאה פעם אחת עבור האיבר שנמצא ב־<var>item</var>.</li>
    <li>הוא חוזר לראש הלולאה שוב, ומנסה לקבל את האיבר הבא באמצעות <code>next</code>. כך עד שייגמרו האיברים ב־generator iterator.</li>
</ol>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    שימו לב שהעובדות שלמדנו בנוגע לאותו "סמן" יבואו לידי ביטוי גם כאן.<br>
    הרצה נוספת של הלולאה על אותו סמן לא תדפיס יותר איברים, כיוון שהסמן מצביע כעת על סוף פונקציית ה־generator:
</p>

In [11]:
for item in our_generator:
    print(item)

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    למזלנו, לולאות <code>for</code> יודעות לטפל בעצמן בשגיאת <code>StopIteration</code>, ולכן שגיאה שכזו לא תקפוץ לנו במקרה הזה.
</p>

#### <span style="text-align: right; direction: rtl; float: right; clear: both;">המרת טיפוסים</span>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    דרך אחרת, לדוגמה, היא לבקש להמיר את ה־generator iterator לסוג משתנה אחר שהוא גם iterable:
</p>

In [12]:
our_generator = silly_generator()
items = list(our_generator)
print(items)

[1, 2, [1, 2, 3]]


<p style="text-align: right; direction: rtl; float: right; clear: both;">
    בקוד שלמעלה, השתמשנו בפונקציה <code>list</code> שיודעת להמיר ערכים iterable־ים לרשימות.<br>
    שימו לב שמה שלמדנו בנוגע ל"סמן" יבוא לידי ביטוי גם בהמרות:
</p>

In [13]:
print(list(our_generator))

[]


### <span style="text-align: right; direction: rtl; float: right; clear: both;">שימושים פרקטיים</span>

#### <span style="text-align: right; direction: rtl; float: right; clear: both;">חיסכון בזיכרון</span>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
נכתוב פונקציה רגילה שמקבלת מספר שלם, ומחזירה רשימה של כל המספרים השלמים מ־0 ועד עד אותו מספר (נשמע לכם מוכר?):
</p>

In [14]:
def my_range(upper_limit):
    numbers = []
    current_number = 0
    while current_number < upper_limit:
        numbers.append(current_number)
        current_number = current_number + 1
    return numbers


for number in my_range(1000):
    print(number)

0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99
100
101
102
103
104
105
106
107
108
109
110
111
112
113
114
115
116
117
118
119
120
121
122
123
124
125
126
127
128
129
130
131
132
133
134
135
136
137
138
139
140
141
142
143
144
145
146
147
148
149
150
151
152
153
154
155
156
157
158
159
160
161
162
163
164
165
166
167
168
169
170
171
172
173
174
175
176
177
178
179
180
181
182
183
184
185
186
187
188
189
190
191
192
193
194
195
196
197
198
199
200
201
202
203
204
205
206
207
208
209
210
211
212
213
214
215
216
217
218
219
220
221
222
223
224
225
226
227
228
229
230
231
232
233
234
235
236
237
238
239
240
241
242
243
244
245
246
247
248
249
250
251
252
253
254
255
256
257
258
259
260
261
262
263
264
265
266
267
268
269
270
271
272
273
274
275
276
27

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    בפונקציה הזו אנחנו יוצרים רשימת מספרים חדשה, המכילה את כל המספרים בין 0 לבין המספר שהועבר לפרמטר <var>upper_limit</var>.<br>
    אך ישנה בעיה חמורה – הפעלת הפונקציה גורמת לניצול משאבים רבים!<br>
    אם נכניס כארגומנט 1,000 – נצטרך להחזיק רשימה המכילה 1,000 איברים שונים, ואם נכניס מספר גדול מדי – עלול להיגמר לנו הזיכרון.
</p>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    אבל איזו סיבה יש לנו להחזיק בזיכרון את רשימת כל המספרים?<br>
    אם לא עולה צורך מובהק שכזה, ייתכן שעדיף להחזיק בזיכרון מספר אחד בלבד בכל פעם, ולהחזירו מייד בעזרת generator:
</p>

In [15]:
def my_range(upper_limit):
    current_number = 0
    while current_number < upper_limit:
        yield current_number
        current_number = current_number + 1


our_generator = my_range(1000)
for number in our_generator:
    print(number)

0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99
100
101
102
103
104
105
106
107
108
109
110
111
112
113
114
115
116
117
118
119
120
121
122
123
124
125
126
127
128
129
130
131
132
133
134
135
136
137
138
139
140
141
142
143
144
145
146
147
148
149
150
151
152
153
154
155
156
157
158
159
160
161
162
163
164
165
166
167
168
169
170
171
172
173
174
175
176
177
178
179
180
181
182
183
184
185
186
187
188
189
190
191
192
193
194
195
196
197
198
199
200
201
202
203
204
205
206
207
208
209
210
211
212
213
214
215
216
217
218
219
220
221
222
223
224
225
226
227
228
229
230
231
232
233
234
235
236
237
238
239
240
241
242
243
244
245
246
247
248
249
250
251
252
253
254
255
256
257
258
259
260
261
262
263
264
265
266
267
268
269
270
271
272
273
274
275
276
27

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    שימו לב כמה הגרסה הזו אלגנטית יותר!<br>
    בכל פעם אנחנו פשוט שולחים את ערכו של מספר אחד (<var>current_number</var>) החוצה.<br>
    כשמבקשים את הערך הבא מה־generator iterator, פונקציית ה־generator חוזרת לעבוד מהנקודה שבה היא עצרה:<br>
    היא מעלה את ערכו של המספר הנוכחי, בודקת אם הוא נמוך מ־<var>upper_limit</var>, ושולחת גם אותו החוצה.<br>
    בשיטה הזו, <code>my_range(numbers)</code> לא מחזירה לנו רשימה של התוצאות – אלא generator iterator שמחזיר ערך אחד בכל פעם.<br>
    כך אנחנו לעולם לא מחזיקים בזיכרון 1,000 מספרים בו־זמנית.
</p>

<div class="align-center" style="display: flex; text-align: right; direction: rtl; clear: both;">
    <div style="display: flex; width: 10%; float: right; clear: both;">
        <img src="images/exercise.svg" style="height: 50px !important;" alt="תרגול"> 
    </div>
    <div style="width: 70%">
        <p style="text-align: right; direction: rtl; float: right; clear: both;">
            לפניכם פונקציה שמקבלת רשימה, ומחזירה עבור כל מספר ברשימה את ערכו בריבוע.<br>
            זוהי גרסה מעט בזבזנית שמשתמשת בהרבה זיכרון. תוכלו להמיר אותה להיות generator?
        </p>
    </div>
    <div style="display: flex; width: 20%; border-right: 0.1rem solid #A5A5A5; padding: 1rem 2rem;">
        <p style="text-align: center; direction: rtl; justify-content: center; align-items: center; clear: both;">
            <strong>חשוב!</strong><br>
            פתרו לפני שתמשיכו!
        </p>
    </div>
</div>

In [4]:
def my_range(upper_limit):
    current_number = 0
    while current_number < upper_limit:
        yield current_number
        current_number = current_number + 1


def square_numbers(numbers):
    for number in numbers:
        yield number**2


our_generator = square_numbers(my_range(1000))
for number in our_generator:
    print(number)

0
1
4
9
16
25
36
49
64
81
100
121
144
169
196
225
256
289
324
361
400
441
484
529
576
625
676
729
784
841
900
961
1024
1089
1156
1225
1296
1369
1444
1521
1600
1681
1764
1849
1936
2025
2116
2209
2304
2401
2500
2601
2704
2809
2916
3025
3136
3249
3364
3481
3600
3721
3844
3969
4096
4225
4356
4489
4624
4761
4900
5041
5184
5329
5476
5625
5776
5929
6084
6241
6400
6561
6724
6889
7056
7225
7396
7569
7744
7921
8100
8281
8464
8649
8836
9025
9216
9409
9604
9801
10000
10201
10404
10609
10816
11025
11236
11449
11664
11881
12100
12321
12544
12769
12996
13225
13456
13689
13924
14161
14400
14641
14884
15129
15376
15625
15876
16129
16384
16641
16900
17161
17424
17689
17956
18225
18496
18769
19044
19321
19600
19881
20164
20449
20736
21025
21316
21609
21904
22201
22500
22801
23104
23409
23716
24025
24336
24649
24964
25281
25600
25921
26244
26569
26896
27225
27556
27889
28224
28561
28900
29241
29584
29929
30276
30625
30976
31329
31684
32041
32400
32761
33124
33489
33856
34225
34596
34969
35344
35721
36100


<p style="text-align: right; direction: rtl; float: right; clear: both;">
    לעיתים ניאלץ לבצע חישוב ארוך, שהשלמתו תימשך זמן רב מאוד.<br>
    במקרה כזה, נוכל להשתמש ב־generator כדי לקבל חלק מהתוצאה בזמן אמת,<br>
    בזמן שבפונקציה "רגילה" נצטרך להמתין עד סיום החישוב כולו.
</p>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    שלשה פיתגורית, לדוגמה, היא שלישיית מספרים שלמים וחיוביים, $a$, $b$ ו־$c$, שעונים על הדרישה $a^2 + b^2 = c^2$.<br>
    אם כך, כדי ששלושה מספרים שאנחנו בוחרים ייחשבו שלשה פיתגורית,<br>
    הסכום של ריבוע המספר הראשון וריבוע המספר השני, אמור להיות שווה לערכו של המספר השלישי בריבוע.
</p>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    אלו דוגמאות לשלשות פיתגוריות:
</p>

<ul style="text-align: right; direction: rtl; float: right; clear: both;">
    <li>$(3, 4, 5)$, כיוון ש־$9 + 16 = 25$.<br>
        9 הוא 3 בריבוע, 16 הוא 4 בריבוע ו־25 הוא 5 בריבוע.
    </li>
    <li>$(5, 12, 13)$, כיוון ש־$25 + 144 = 169$.</li>
    <li>$(8, 15, 17)$, כיוון ש־$64 + 225 = 289$.</li>
</ul>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    ננסה למצוא את כל השלשות הפיתגוריות מתחת ל־10,000 בעזרת קוד שרץ על כל השלשות האפשריות:
</p>

In [ ]:
def find_pythagorean_triples(upper_bound=10_000):
    pythagorean_triples = []
    for a in range(2, upper_bound):
        for b in range(1, a):
            for c in range(1, b):
                if a ** 2 + b **2 == c ** 2:
                    pythagorean_triples.append((a, b, c))
    return pythagorean_triples


for triple in find_pythagorean_triples():
    print(triple)

<div class="align-center" style="display: flex; text-align: right; direction: rtl;">
    <div style="display: flex; width: 10%; float: right; ">
        <img src="images/warning.png" style="height: 50px !important;" alt="אזהרה!"> 
    </div>
    <div style="width: 90%">
        <p style="text-align: right; direction: rtl;">
            הרצת התא הקודם תתקע את המחברת (חישוב התוצאה יימשך זמן רב).<br>
            כדי להיות מסוגלים להריץ את התאים הבאים, לחצו <samp>00</samp> לאחר הרצת התא, ובחרו <em>Restart</em>.<br>
            אל דאגה – האתחול יתבצע אך ורק עבור המחברת, ולא עבור מחשב.
        </p>
    </div>
</div>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    יו, כמה זמן נמשכת הרצת הקוד הזה... 😴<br>
    הלוואי שעד שהקוד הזה היה מסיים היינו מקבלים לפחות <em>חלק</em> מהתוצאות!<br>
    נפנה ל־generator־ים לעזרה:
</p>

In [ ]:
def find_pythagorean_triples(upper_bound=10_000):
    for c in range(3, upper_bound):
        for b in range(2, c):
            for a in range(1, b):
                if a ** 2 + b **2 == c ** 2:
                    yield a, b, c


for triple in find_pythagorean_triples():
    print(triple)

(3, 4, 5)
(6, 8, 10)
(5, 12, 13)
(9, 12, 15)
(8, 15, 17)
(12, 16, 20)
(15, 20, 25)
(7, 24, 25)
(10, 24, 26)
(20, 21, 29)
(18, 24, 30)
(16, 30, 34)
(21, 28, 35)
(12, 35, 37)
(15, 36, 39)
(24, 32, 40)
(9, 40, 41)
(27, 36, 45)
(30, 40, 50)
(14, 48, 50)
(24, 45, 51)
(20, 48, 52)
(28, 45, 53)
(33, 44, 55)
(40, 42, 58)
(36, 48, 60)
(11, 60, 61)
(39, 52, 65)
(33, 56, 65)
(25, 60, 65)
(16, 63, 65)
(32, 60, 68)
(42, 56, 70)
(48, 55, 73)
(24, 70, 74)
(45, 60, 75)
(21, 72, 75)
(30, 72, 78)
(48, 64, 80)
(18, 80, 82)
(51, 68, 85)
(40, 75, 85)
(36, 77, 85)
(13, 84, 85)
(60, 63, 87)
(39, 80, 89)
(54, 72, 90)
(35, 84, 91)
(57, 76, 95)
(65, 72, 97)
(60, 80, 100)
(28, 96, 100)
(20, 99, 101)
(48, 90, 102)
(40, 96, 104)
(63, 84, 105)
(56, 90, 106)
(60, 91, 109)
(66, 88, 110)
(36, 105, 111)
(15, 112, 113)
(69, 92, 115)
(80, 84, 116)
(45, 108, 117)
(56, 105, 119)
(72, 96, 120)
(22, 120, 122)
(27, 120, 123)
(75, 100, 125)
(44, 117, 125)
(35, 120, 125)
(78, 104, 130)
(66, 112, 130)
(50, 120, 130)
(32, 126, 13

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    איך זה קרה? קיבלנו את התשובה בתוך שבריר שנייה!<br>
    ובכן, זה לא מדויק – קיבלנו חלק מהתשובות. שימו לב שהקוד ממשיך להדפיס :)<br>
</p>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    להזכירכם, ה־generator שולח את התוצאה החוצה מייד כשהוא מוצא שלשה אחת,<br>
    וה־for מקבל מה־generator iterable כל שלשה ברגע שהיא נמצאה.<br>
    ברגע שה־for מקבל שלשה, הוא מבצע את גוף הלולאה עבור אותה שלשה, ורק אז מבקש מ־generator את האיבר הבא.<br>
    בגלל האופי של generators, הקוד בתא האחרון מדפיס לנו כל שלשה ברגע שהוא מצא אותה, ולא מחכה עד שיימצאו כל השלשות.
</p>

#### <span style="text-align: right; direction: rtl; float: right; clear: both;">תרגול ביניים: מספרים פראיים</span>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    "פירוק לגורמים של מספר שלם" היא בעיה שחישוב פתרונה נמשך זמן רב במחשבים מודרניים.
</p>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    עליכם לכתוב פונקציה שמקבלת מספר חיובי שלם $n$, ומחזירה קבוצת מספרים שמכפלתם (תוצאת הכפל ביניהם) היא $n$.<br>
    לדוגמה, המספר 1,386 בנוי מהמכפלה של קבוצת המספרים $2 \cdot 3 \cdot 3 \cdot 7 \cdot 11$.<br>
    כל מספר בקבוצת המספרים הזו חייב להיות ראשוני.<br>
    להזכירכם: מספר ראשוני הוא מספר שאין לו מחלקים חוץ מעצמו ומ־1.
</p>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    הניחו שהמספר שהתקבל אינו ראשוני.<br>
    מה היתרון של generator על פני פונקציה רגילה שעושה אותו דבר?
</p>

<p style="text-align: right; direction: rtl; float: right; clear: both;">    
    רמז: <span style="background: black;">אם תנסו לחלק את המספר ב־2, ואז ב־3 (וכן הלאה), בסופו של דבר תגיעו למחלק ראשוני של המספר.</span><br>
    רמז עבה: <span style="background: black;">בכל פעם שמצאתם מחלק אחד למספר, חלקו את המספר במחלק, והתחילו את החיפוש מחדש. מתי עליכם לעצור?</span>
</p>

In [16]:
# Upload 83
def factorization_func(n):
    list_of_divides = []
    divider = 2
    while n != 1:
        if n % divider == 0:
            list_of_divides.append(divider)
            n = n / divider
            divider = 1
        divider += 1
    return list_of_divides


print(factorization_func(999999999999999999))

# another option with generator
def factorization_generator(n):
    divider = 2
    while n != 1:
        if n % divider == 0:
            yield divider
            n = n / divider
            divider = 1
        divider += 1


my_generator = factorization_generator(1386)
for divider in my_generator:
    print(divider)

[3, 2, 2, 2, 2, 2, 2, 3, 7, 109, 109, 167, 499, 250501]
2
3
3
7
11


2
3
3
7
11


#### <span style="text-align: right; direction: rtl; float: right; clear: both;">אוספים אין־סופיים</span>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    עבור בעיות מסוימות, נרצה להיות מסוגלים להחזיר אין־סוף תוצאות.<br>
    ניקח כדוגמה לסדרה אין־סופית את סדרת פיבונאצ'י, שבה כל איבר הוא סכום זוג האיברים הקודמים לו:<br>
    $1, 1, 2, 3, 5, 8, \ldots$
</p>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    נממש פונקציה שמחזירה לנו את סדרת פיבונאצ'י.<br>
    בפונקציה רגילה אין לנו אפשרות להחזיר מספר אין־סופי של איברים, ולכן נצטרך להחליט על מספר האיברים המרבי שנרצה להחזיר:
</p>

In [174]:
def fibonacci(max_items):
    a = 1
    b = 1
    numbers = [1, 1]
    while len(numbers) < max_items:
        a, b = b, a + b  # Unpacking
        numbers.append(b)
    return numbers


for number in fibonacci(10):
    print(number)

1
1
2
3
5
8
13
21
34
55


<p style="text-align: right; direction: rtl; float: right; clear: both;">
    לעומת זאת, ל־generators לא חייב להיות סוף מוגדר.<br>
    נשתמש ב־<code>while True</code> שתמיד מתקיים, כדי שבסופו של דבר – תמיד נגיע ל־<code>yield</code>: 
</p>

In [9]:
def fibonacci():
    a = 1
    b = 1
    numbers = [1, 1]
    while True:  # תמיד מתקיים
        yield a
        a, b = b, a + b

        
generator_iterator = fibonacci()
for number in range(10):
    print(next(generator_iterator))

# אני יכול לבקש בקלות רבה את 10 האיברים הבאים בסדרה
for number in range(10):
    print(next(generator_iterator))

1
1
2
3
5
8
13
21
34
55
89
144
233
377
610
987
1597
2584
4181
6765


<div class="align-center" style="display: flex; text-align: right; direction: rtl;">
    <div style="display: flex; width: 10%; float: right; ">
        <img src="images/warning.png" style="height: 50px !important;" alt="אזהרה!"> 
    </div>
    <div style="width: 90%">
        <p style="text-align: right; direction: rtl;">
            generators אין־סופיים יכולים לגרום בקלות ללולאות אין־סופיות, גם בלולאות <code>for</code>.<br>
            שימו לב לצורת ההתעסקות העדינה בדוגמאות למעלה.<br>
            הרצת לולאת <code>for</code> ישירות על ה־generator iterator הייתה מכניסה אותנו ללולאה אין־סופית.
        </p>
    </div>
</div>

<div class="align-center" style="display: flex; text-align: right; direction: rtl; clear: both;">
    <div style="display: flex; width: 10%; float: right; clear: both;">
        <img src="images/exercise.svg" style="height: 50px !important;" alt="תרגול"> 
    </div>
    <div style="width: 70%">
        <p style="text-align: right; direction: rtl; float: right; clear: both;">
            כתבו generator שמחזיר את כל המספרים השלמים הגדולים מ־0.
        </p>
    </div>
    <div style="display: flex; width: 20%; border-right: 0.1rem solid #A5A5A5; padding: 1rem 2rem;">
        <p style="text-align: center; direction: rtl; justify-content: center; align-items: center; clear: both;">
            <strong>חשוב!</strong><br>
            פתרו לפני שתמשיכו!
        </p>
    </div>
</div>

In [11]:
def integer_numbers_are_greater_than_zero():
    number = 0
    while True:  # תמיד מתקיים
        yield number
        number += 1


generator_iterator = integer_numbers_are_greater_than_zero()
for number in range(100):
    print(next(generator_iterator))

0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99


### <span style="text-align: right; direction: rtl; float: right; clear: both;">ריבוי generator iterators</span>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    נגדיר generator פשוט שמחזיר את האיברים <samp>1</samp>, <samp>2</samp> ו־<samp>3</samp>:
</p>

In [12]:
def simple_generator():
    yield 1
    yield 2
    yield 3

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    ניצור שני generator iterators ("סמנים") שונים שמצביעים לשורה הראשונה של ה־generator שמופיע למעלה:
</p>

In [13]:
first_gen = simple_generator()
second_gen = simple_generator()

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    בעניין זה, חשוב להבין שכל אחד מה־generator iterators הוא "חץ" נפרד שמצביע לשורה הראשונה ב־<var>simple_generator</var>.<br>
    אם נבקש מכל אחד מהם להחזיר ערך, נקבל משניהם את 1, ואותו חץ דמיוני יעבור בשני ה־generator iterators להמתין בשורה השנייה:
</p>

In [14]:
print(next(first_gen))
print(next(second_gen))

1
1


<p style="text-align: right; direction: rtl; float: right; clear: both;">
    נוכל לקדם את <var>first_gen</var>, לדוגמה, לסוף הפונקציה:
</p>

In [15]:
print(next(first_gen))
print(next(first_gen))

2
3


<p style="text-align: right; direction: rtl; float: right; clear: both;">
    אבל <var>second_gen</var> הוא חץ נפרד, שעדיין מצביע לשורה השנייה של פונקציית ה־generator.<br>
    אם נבקש ממנו את הערך הבא, הוא ימשיך את המסע מהערך <samp>2</samp>:<br>
</p>

In [16]:
print(next(second_gen))

2


<p style="text-align: right; direction: rtl; float: right; clear: both;">
    ממצב זה נוכל להסיק שאפשר ליצור יותר מ־generator iterator אחד עבור כל פונקציית generator.<br>
    כל אחד מה־generator iterators יחזיק תמונת מצב עצמאית של המקום שבו עצר הסמן ושל ערכי המשתנים.<br>
    ההתנהלות של כל generator iterator תקרה בנפרד, ולא תושפע בשום צורה מה־generator iterators האחרים.<br>
</p>

### <span style="text-align: right; direction: rtl; float: right; clear: both;">הבדלי מינוח</span>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    בשלב זה יש לנו הרבה מבנים שאפשר לרוץ עליהם, וכל המינוח סביב עניין האיטרביליות נעשה מעט מבלבל.<br>
    ננסה לעשות סדר בדברים:
</p>

<dl style="text-align: right; direction: rtl; float: right; clear: both;">
<dt>Iterable</dt><dd>
    אם ערך מסוים הוא iterable, אפשר לפרק אותו ליחידות קטנות יותר, ולהתייחס לכל יחידה בנפרד.
</dd>
<dt>Iteration, חִזְרוּר</dt><dd>ביצוע יחיד של גוף הלולאה עבור ערך מסוים.</dd>
<dt>Iterator</dt><dd>
    ערך שמייצג זרם של מידע, ומתוכו מאחזרים ערכים אחרים. אפשר לאחזר ממנו ערך אחד בכל פעם, לפי סדר מסוים, בעזרת <code dir="ltr">next()</code>.<br>
    iterator הוא בהכרח iterable, אך לא כל iterable הוא iterator.
</dd>
<dt>Sequence</dt>
    <dd>
        כל iterable שאפשר לחלץ ממנו איברים באמצעות פנייה למיקום שלהם (<code>iterable[0]</code>), כמו מחרוזות, רשימות ו־tuple־ים.<br>
        sequence הוא בהכרח iterable, אך לא כל iterable הוא sequence.
    </dd>
<dt>פונקציית ה־generator</dt><dd>פונקציה המכילה <code>yield</code> ומגדירה אילו ערכים יוחזרו מה־generator.</dd>
<dt>Generator iterator</dt>
    <dd>
        iterator שנוצר מתוך פונקציית ה־generator.
    </dd>
<dt>Generator</dt><dd>לרוב מתייחס לפונקציית ה־generator, אך יש פעמים שמשתמשים במינוח כדי להתייחס ל־generator iterator.</dd>
</dl>

## <span style="text-align: right; direction: rtl; float: right; clear: both;">סיכום</span>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    Generators הם פונקציות שמאפשרות לנו להחזיר סדרות ערכים באופן מדורג.<br>
    כשנקרא לפונקציית generator, היא תחזיר לנו generator iterator שישמש מעין "סמן".<br>
    הסמן ישמור "מצב" שמתאר את המקום שבו אנחנו שוהים בתוך הפונקציה, ואת הערכים שחושבו במהלך ריצתה עד כה.<br>
    בכל שלב, נוכל לבקש את הערך הבא ב־generator בעזרת קריאה לפונקציה <code>next</code> על ה־generator iterator.<br>
    נוכל גם להשתמש במבנים שיזיזו עבורנו את הסמן, כמו <code>for</code> או המרה לסוג נתונים אחר שגם הוא iterable.
</p>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    ל־generators יתרונות רבים:
</p>

<ul style="text-align: right; direction: rtl; float: right; clear: both;">
    <li>אפשר ליצור בעזרתם פונקציות שמחזירות מספר אין־סופי של נתונים.</li>
    <li>במקרים מסוימים, נוכל להיעזר בהם כדי לקבל רק חלק מהתוצאות בכל זמן נתון.</li>
    <li>שימוש נכון בהם יכול להיות מפתח לחיסכון משמעותי במשאבי התוכנית.</li>
</ul>

## <span style="text-align: right; direction: rtl; float: right; clear: both;">תרגילים</span>

### <span style="text-align: right; direction: rtl; float: right; clear: both;">פיצוץ אוכלוסין</span>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    קראו בוויקיפדיה על דרך החישוב של <a href="https://he.wikipedia.org/wiki/%D7%A1%D7%A4%D7%A8%D7%AA_%D7%91%D7%99%D7%A7%D7%95%D7%A8%D7%AA#%D7%A1%D7%A4%D7%A8%D7%AA_%D7%91%D7%99%D7%A7%D7%95%D7%A8%D7%AA_%D7%91%D7%9E%D7%A1%D7%A4%D7%A8_%D7%94%D7%96%D7%94%D7%95%D7%AA_%D7%91%D7%99%D7%A9%D7%A8%D7%90%D7%9C">ספרת הביקורת</a> במספרי הזהות בישראל. 
</p>

<ol style="text-align: right; direction: rtl; float: right; clear: both;">
    <li>ממשו פונקציה שמקבלת מספר זהות ללא ספרת ביקורת, ומחזירה את ספרת הביקורת.</li>
    <li>ממשו תוכנית המדפיסה את כל מספרי הזהות האפשריים במדינת ישראל. השתמשו בקוד שכתבתם בסעיף הראשון.</li>
</ol>

In [18]:
# Upload 84
import math

def round_up(n, decimals=0):
    multiplier = 10 ** decimals
    return math.ceil(n * multiplier) / multiplier


def check_digit(user_id):
    flag_weight = 2
    total_calc = 0
    while user_id:
        number = user_id % 10
        if flag_weight == 2:
            number = (number * flag_weight)
            flag_weight = 1
        else:
            number = (number * flag_weight)
            flag_weight = 2
        if number >= 10:
            total_calc += number % 10 + 1
        else:
            total_calc += number
        user_id = user_id // 10
    rounded_number = round_up(total_calc, decimals=-1)
    return rounded_number - total_calc


def all_identity_numbers():
   # all_id = 312162951
    all_id = 0
    while all_id < 99999999:
        real_id = all_id * 10 + check_digit(all_id)
        real_id = str(real_id)
        real_id = real_id[:-2]
        yield real_id.zfill(9)
        all_id += 1


my_genrator = all_identity_numbers()
for number in range(25000):
    print(next(my_genrator))

000000000
000000018
000000026
000000034
000000042
000000059
000000067
000000075
000000083
000000091
000000109
000000117
000000125
000000133
000000141
000000158
000000166
000000174
000000182
000000190
000000208
000000216
000000224
000000232
000000240
000000257
000000265
000000273
000000281
000000299
000000307
000000315
000000323
000000331
000000349
000000356
000000364
000000372
000000380
000000398
000000406
000000414
000000422
000000430
000000448
000000455
000000463
000000471
000000489
000000497
000000505
000000513
000000521
000000539
000000547
000000554
000000562
000000570
000000588
000000596
000000604
000000612
000000620
000000638
000000646
000000653
000000661
000000679
000000687
000000695
000000703
000000711
000000729
000000737
000000745
000000752
000000760
000000778
000000786
000000794
000000802
000000810
000000828
000000836
000000844
000000851
000000869
000000877
000000885
000000893
000000901
000000919
000000927
000000935
000000943
000000950
000000968
000000976
000000984
000000992


000017541
000017558
000017566
000017574
000017582
000017590
000017608
000017616
000017624
000017632
000017640
000017657
000017665
000017673
000017681
000017699
000017707
000017715
000017723
000017731
000017749
000017756
000017764
000017772
000017780
000017798
000017806
000017814
000017822
000017830
000017848
000017855
000017863
000017871
000017889
000017897
000017905
000017913
000017921
000017939
000017947
000017954
000017962
000017970
000017988
000017996
000018002
000018010
000018028
000018036
000018044
000018051
000018069
000018077
000018085
000018093
000018101
000018119
000018127
000018135
000018143
000018150
000018168
000018176
000018184
000018192
000018200
000018218
000018226
000018234
000018242
000018259
000018267
000018275
000018283
000018291
000018309
000018317
000018325
000018333
000018341
000018358
000018366
000018374
000018382
000018390
000018408
000018416
000018424
000018432
000018440
000018457
000018465
000018473
000018481
000018499
000018507
000018515
000018523
000018531


000041129
000041137
000041145
000041152
000041160
000041178
000041186
000041194
000041202
000041210
000041228
000041236
000041244
000041251
000041269
000041277
000041285
000041293
000041301
000041319
000041327
000041335
000041343
000041350
000041368
000041376
000041384
000041392
000041400
000041418
000041426
000041434
000041442
000041459
000041467
000041475
000041483
000041491
000041509
000041517
000041525
000041533
000041541
000041558
000041566
000041574
000041582
000041590
000041608
000041616
000041624
000041632
000041640
000041657
000041665
000041673
000041681
000041699
000041707
000041715
000041723
000041731
000041749
000041756
000041764
000041772
000041780
000041798
000041806
000041814
000041822
000041830
000041848
000041855
000041863
000041871
000041889
000041897
000041905
000041913
000041921
000041939
000041947
000041954
000041962
000041970
000041988
000041996
000042002
000042010
000042028
000042036
000042044
000042051
000042069
000042077
000042085
000042093
000042101
000042119


000064717
000064725
000064733
000064741
000064758
000064766
000064774
000064782
000064790
000064808
000064816
000064824
000064832
000064840
000064857
000064865
000064873
000064881
000064899
000064907
000064915
000064923
000064931
000064949
000064956
000064964
000064972
000064980
000064998
000065003
000065011
000065029
000065037
000065045
000065052
000065060
000065078
000065086
000065094
000065102
000065110
000065128
000065136
000065144
000065151
000065169
000065177
000065185
000065193
000065201
000065219
000065227
000065235
000065243
000065250
000065268
000065276
000065284
000065292
000065300
000065318
000065326
000065334
000065342
000065359
000065367
000065375
000065383
000065391
000065409
000065417
000065425
000065433
000065441
000065458
000065466
000065474
000065482
000065490
000065508
000065516
000065524
000065532
000065540
000065557
000065565
000065573
000065581
000065599
000065607
000065615
000065623
000065631
000065649
000065656
000065664
000065672
000065680
000065698
000065706


000089755
000089763
000089771
000089789
000089797
000089805
000089813
000089821
000089839
000089847
000089854
000089862
000089870
000089888
000089896
000089904
000089912
000089920
000089938
000089946
000089953
000089961
000089979
000089987
000089995
000090001
000090019
000090027
000090035
000090043
000090050
000090068
000090076
000090084
000090092
000090100
000090118
000090126
000090134
000090142
000090159
000090167
000090175
000090183
000090191
000090209
000090217
000090225
000090233
000090241
000090258
000090266
000090274
000090282
000090290
000090308
000090316
000090324
000090332
000090340
000090357
000090365
000090373
000090381
000090399
000090407
000090415
000090423
000090431
000090449
000090456
000090464
000090472
000090480
000090498
000090506
000090514
000090522
000090530
000090548
000090555
000090563
000090571
000090589
000090597
000090605
000090613
000090621
000090639
000090647
000090654
000090662
000090670
000090688
000090696
000090704
000090712
000090720
000090738
000090746


000114462
000114470
000114488
000114496
000114504
000114512
000114520
000114538
000114546
000114553
000114561
000114579
000114587
000114595
000114603
000114611
000114629
000114637
000114645
000114652
000114660
000114678
000114686
000114694
000114702
000114710
000114728
000114736
000114744
000114751
000114769
000114777
000114785
000114793
000114801
000114819
000114827
000114835
000114843
000114850
000114868
000114876
000114884
000114892
000114900
000114918
000114926
000114934
000114942
000114959
000114967
000114975
000114983
000114991
000115006
000115014
000115022
000115030
000115048
000115055
000115063
000115071
000115089
000115097
000115105
000115113
000115121
000115139
000115147
000115154
000115162
000115170
000115188
000115196
000115204
000115212
000115220
000115238
000115246
000115253
000115261
000115279
000115287
000115295
000115303
000115311
000115329
000115337
000115345
000115352
000115360
000115378
000115386
000115394
000115402
000115410
000115428
000115436
000115444
000115451


000135707
000135715
000135723
000135731
000135749
000135756
000135764
000135772
000135780
000135798
000135806
000135814
000135822
000135830
000135848
000135855
000135863
000135871
000135889
000135897
000135905
000135913
000135921
000135939
000135947
000135954
000135962
000135970
000135988
000135996
000136002
000136010
000136028
000136036
000136044
000136051
000136069
000136077
000136085
000136093
000136101
000136119
000136127
000136135
000136143
000136150
000136168
000136176
000136184
000136192
000136200
000136218
000136226
000136234
000136242
000136259
000136267
000136275
000136283
000136291
000136309
000136317
000136325
000136333
000136341
000136358
000136366
000136374
000136382
000136390
000136408
000136416
000136424
000136432
000136440
000136457
000136465
000136473
000136481
000136499
000136507
000136515
000136523
000136531
000136549
000136556
000136564
000136572
000136580
000136598
000136606
000136614
000136622
000136630
000136648
000136655
000136663
000136671
000136689
000136697


000158113
000158121
000158139
000158147
000158154
000158162
000158170
000158188
000158196
000158204
000158212
000158220
000158238
000158246
000158253
000158261
000158279
000158287
000158295
000158303
000158311
000158329
000158337
000158345
000158352
000158360
000158378
000158386
000158394
000158402
000158410
000158428
000158436
000158444
000158451
000158469
000158477
000158485
000158493
000158501
000158519
000158527
000158535
000158543
000158550
000158568
000158576
000158584
000158592
000158600
000158618
000158626
000158634
000158642
000158659
000158667
000158675
000158683
000158691
000158709
000158717
000158725
000158733
000158741
000158758
000158766
000158774
000158782
000158790
000158808
000158816
000158824
000158832
000158840
000158857
000158865
000158873
000158881
000158899
000158907
000158915
000158923
000158931
000158949
000158956
000158964
000158972
000158980
000158998
000159004
000159012
000159020
000159038
000159046
000159053
000159061
000159079
000159087
000159095
000159103


000182907
000182915
000182923
000182931
000182949
000182956
000182964
000182972
000182980
000182998
000183004
000183012
000183020
000183038
000183046
000183053
000183061
000183079
000183087
000183095
000183103
000183111
000183129
000183137
000183145
000183152
000183160
000183178
000183186
000183194
000183202
000183210
000183228
000183236
000183244
000183251
000183269
000183277
000183285
000183293
000183301
000183319
000183327
000183335
000183343
000183350
000183368
000183376
000183384
000183392
000183400
000183418
000183426
000183434
000183442
000183459
000183467
000183475
000183483
000183491
000183509
000183517
000183525
000183533
000183541
000183558
000183566
000183574
000183582
000183590
000183608
000183616
000183624
000183632
000183640
000183657
000183665
000183673
000183681
000183699
000183707
000183715
000183723
000183731
000183749
000183756
000183764
000183772
000183780
000183798
000183806
000183814
000183822
000183830
000183848
000183855
000183863
000183871
000183889
000183897


000209122
000209130
000209148
000209155
000209163
000209171
000209189
000209197
000209205
000209213
000209221
000209239
000209247
000209254
000209262
000209270
000209288
000209296
000209304
000209312
000209320
000209338
000209346
000209353
000209361
000209379
000209387
000209395
000209403
000209411
000209429
000209437
000209445
000209452
000209460
000209478
000209486
000209494
000209502
000209510
000209528
000209536
000209544
000209551
000209569
000209577
000209585
000209593
000209601
000209619
000209627
000209635
000209643
000209650
000209668
000209676
000209684
000209692
000209700
000209718
000209726
000209734
000209742
000209759
000209767
000209775
000209783
000209791
000209809
000209817
000209825
000209833
000209841
000209858
000209866
000209874
000209882
000209890
000209908
000209916
000209924
000209932
000209940
000209957
000209965
000209973
000209981
000209999
000210005
000210013
000210021
000210039
000210047
000210054
000210062
000210070
000210088
000210096
000210104
000210112


000236752
000236760
000236778
000236786
000236794
000236802
000236810
000236828
000236836
000236844
000236851
000236869
000236877
000236885
000236893
000236901
000236919
000236927
000236935
000236943
000236950
000236968
000236976
000236984
000236992
000237008
000237016
000237024
000237032
000237040
000237057
000237065
000237073
000237081
000237099
000237107
000237115
000237123
000237131
000237149
000237156
000237164
000237172
000237180
000237198
000237206
000237214
000237222
000237230
000237248
000237255
000237263
000237271
000237289
000237297
000237305
000237313
000237321
000237339
000237347
000237354
000237362
000237370
000237388
000237396
000237404
000237412
000237420
000237438
000237446
000237453
000237461
000237479
000237487
000237495
000237503
000237511
000237529
000237537
000237545
000237552
000237560
000237578
000237586
000237594
000237602
000237610
000237628
000237636
000237644
000237651
000237669
000237677
000237685
000237693
000237701
000237719
000237727
000237735
000237743


In [152]:
def round_up(n, decimals=0):
    multiplier = 10 ** decimals
    return math.ceil(n * multiplier) / multiplier

round_up(11,-1)

20.0

### <span style="text-align: right; direction: rtl; float: right; clear: both;">מנה מושלמת לחלוקה</span>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    אפשר לחלק רול סושי של 6 יחידות לאדם אחד, ל־2 אנשים, ל־3 אנשים ול־6 אנשים.<br>
    נתעלם ממצבים שבהם כל אדם מקבל רק חתיכת סושי אחת. זה נשמע לי עצוב.<br>
    נגדיר "מנה מושלמת לחלוקה" כמנה שאם נסכום את כל הצורות לחלק אותה, נקבל את גודל המנה עצמה.
</p>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    לדוגמה:
</p>

<ul style="text-align: right; direction: rtl; float: right; clear: both;">
    <li>רול סושי בעל 6 יחידות הוא מנה מושלמת לחלוקה, כיוון שאפשר לחלק אותו לאדם 1, ל־2 אנשים או ל־3 אנשים. $1+2+3=6$.</li>
    <li>רול סושי בעל 8 יחידות הוא לא מנה מושלמת לחלוקה, כי אפשר לחלק אותו לאדם 1, ל־2 אנשים או ל־4 אנשים. $1+2+4 \neq 8$.</li>
    <li>רול בעל 12 יחידות גם הוא לא מנה מושלמת לחלוקה – $1 + 2 + 3 + 4 + 6 \neq 12$.</li>
    <li>רול בעל 28 יחידות הוא בהחלט מנה מושלמת לחלוקה – $1 + 2 + 4 + 7 + 14 = 28$.</li>
</ul>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    כתבו תוכנית שמדפיסה באופן אין־סופי את כל גודלי המנות שנחשבים מושלמים לחלוקה.
</p>

In [8]:
def fibonacci():
    a = 1
    b = 1
    numbers = [1, 1]
    while True:  # תמיד מתקיים
        yield a
        a, b = b, a + b

        
generator_iterator = fibonacci()
for number in range(10):
    print(next(generator_iterator))

# אני יכול לבקש בקלות רבה את 10 האיברים הבאים בסדרה
for number in range(10):
    print(next(generator_iterator))

1
1
2
3
5
8
13
21
34
55
89
144
233
377
610
987
1597
2584
4181
6765


In [21]:
# Upload 85
def perfect_dish_for_distribution():
    number = 2
    i = 1
    divider = 0
    while True:
        while i < number:
            if number % i == 0:
                divider += i
            i += 1
        if divider == number:
            yield number
        number += 1
        divider = 0
        i = 1


my_genrator = perfect_dish_for_distribution()   
for divider in range(5):
    # print(divider)
    print(next(my_genrator))
# another option different implemention
def perfect_dish_for_distribution(number):
    divider = 0
    i = 1
    while i < number:
        if number % i == 0:
            yield i
        i += 1


def check_perfect_dish(dividers,number):
    if dividers == number:
        return True
    return False


number = int(input("please enter a number of pieces from sushi: "))
my_genrator = perfect_dish_for_distribution(number)
dividers = 0
for divider in my_genrator:
    print(divider)
    dividers += divider
if check_perfect_dish(dividers,number):
    print("This is perfect dish for distribution")
else:
    print("This isn't perfect dish for distribution")

6
28
496
8128


KeyboardInterrupt: 

In [28]:
##### fixxx ########
def perfect_dish_for_distribution(number):
    divider = 0
    i = 1
    while i < number:
        if number % i == 0:
            yield i
        i += 1


def check_perfect_dish(dividers,number):
    if dividers == number:
        return True
    return False


number = int(input("please enter a number of pieces from sushi: "))
my_genrator = perfect_dish_for_distribution(number)
dividers = 0
for divider in my_genrator:
    print(divider)
    dividers += divider
if check_perfect_dish(dividers,number):
    print("This is perfect dish for distribution")
else:
    print("This isn't perfect dish for distribution")

please enter a number of pieces from sushi: 8589869056
1
2
4
8
16
32
64
128
256
512
1024
2048
4096
8192
16384
32768
65536
131071
262142
524284
1048568
2097136
4194272
8388544
16777088
33554176
67108352
134216704
268433408
536866816
1073733632


KeyboardInterrupt: 

In [108]:
# ##### fixxx ########
# def perfect_dish_for_distribution():
#     divider = 2
#     list_of_divider = []
#     i = 1
#     while i < divider:
#         if divider % i == 0:
#             yield i
#         i += 1


# my_genrator = perfect_dish_for_distribution()
# for divider in range(1):
#     print(next(my_genrator))


1


In [22]:

def perfect_dish_for_distribution(number):
    divider = 0
    i = 1
    while i < number:
        if number % i == 0:
            yield i
        i += 1


def check_perfect_dish(dividers,number):
    if dividers == number:
        return True
    return False



dividers = 0
for divider in range(1, 1000):
    print(dividers)
   # dividers = 0
    my_genrator = perfect_dish_for_distribution(divider)
    dividers += divider
    if check_perfect_dish(dividers,divider):
        print(f"The number {divider} This is perfect dish for distribution")
    else:
        print(f"The number {divider} This isn't perfect dish for distribution")

0
The number 1 This is perfect dish for distribution
1
The number 2 This isn't perfect dish for distribution
3
The number 3 This isn't perfect dish for distribution
6
The number 4 This isn't perfect dish for distribution
10
The number 5 This isn't perfect dish for distribution
15
The number 6 This isn't perfect dish for distribution
21
The number 7 This isn't perfect dish for distribution
28
The number 8 This isn't perfect dish for distribution
36
The number 9 This isn't perfect dish for distribution
45
The number 10 This isn't perfect dish for distribution
55
The number 11 This isn't perfect dish for distribution
66
The number 12 This isn't perfect dish for distribution
78
The number 13 This isn't perfect dish for distribution
91
The number 14 This isn't perfect dish for distribution
105
The number 15 This isn't perfect dish for distribution
120
The number 16 This isn't perfect dish for distribution
136
The number 17 This isn't perfect dish for distribution
153
The number 18 This isn'

477753
The number 978 This isn't perfect dish for distribution
478731
The number 979 This isn't perfect dish for distribution
479710
The number 980 This isn't perfect dish for distribution
480690
The number 981 This isn't perfect dish for distribution
481671
The number 982 This isn't perfect dish for distribution
482653
The number 983 This isn't perfect dish for distribution
483636
The number 984 This isn't perfect dish for distribution
484620
The number 985 This isn't perfect dish for distribution
485605
The number 986 This isn't perfect dish for distribution
486591
The number 987 This isn't perfect dish for distribution
487578
The number 988 This isn't perfect dish for distribution
488566
The number 989 This isn't perfect dish for distribution
489555
The number 990 This isn't perfect dish for distribution
490545
The number 991 This isn't perfect dish for distribution
491536
The number 992 This isn't perfect dish for distribution
492528
The number 993 This isn't perfect dish for distr

### <span style="text-align: right; direction: rtl; float: right; clear: both;">לחששנית</span>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    בקובץ resources/logo.jpg מופיע לוגו הקורס, ובתוכו מוכמנים מסרים סודיים אחדים.<br>
    המסרים הם מחרוזות באורך 5 אותיות לפחות, כתובים באותיות אנגליות קטנות בלבד ומסתיימים בסימן קריאה.
</p>

<p style="text-align: right; direction: rtl; float: right; clear: both;">
    פתחו את הלוגו לקריאה בתצורה בינארית, וחלצו ממנו את המסרים הסודיים.<br>
    זכרו שהקובץ עלול להיות גדול מאוד, ועדיף שלא לקרוא את כולו במכה אחת.<br>
    מצאו באינטרנט עזרה בנוגע לפתיחת קבצים בצורה בינארית ולקריאה מדורגת של הקובץ.<br>
    הקפידו שלא להשתמש בטכניקות שלא למדנו (או להוסיף אותן רק בנוסף לפתרון שכזה).
</p>

In [3]:
import re
with open('resources/logo.jpg', 'rb') as file_binary:
       print re.findall("([a-z]{5,})",  file_binary.read())

TypeError: cannot use a string pattern on a bytes-like object

In [31]:
# Upload 72
def check_if_the_msg_in_line(line):
    msg = ''
    flag = 1
    line = line.split('\\')
    for word in line:
        if '!' in word:
            flag = 1
            msg = ''
            index = word.index('!') - 1
            while flag:
                temp = word[index]
                if (temp.islower() == True) :
                    msg += word[index]
                else:
                    length_msg = len(msg)
                    if msg == '' or length_msg < 5 :
                        flag = 0
                    else:
                        flag = 0
                        print(msg[::-1])
                index -= 1


with open('resources/logo.jpg', 'rb') as file_binary:
    text_binary = file_binary.readlines()

i = 0
for line in text_binary:
    line = str(line)
    check_if_the_msg_in_line(line)


# python isawesome welldone goodjob

python
isawesome
welldone
goodjob


In [103]:
def check_if_the_msg_in_line():
    msg = ''
    flag = 1
    line = line.split('\\')
    for word in line:
        if '!' in word:
            flag = 1
            msg = ''
            index = word.index('!') - 1
            while flag:
                temp = word[index]
                if (temp.islower() == True) :
                    msg += word[index]
                else:
                    length_msg = len(msg)
                    if msg == '' or length_msg < 5 :
                        flag = 0
                    else:
                        flag = 0
                        yield(msg[::-1])
                index -= 1


with open('resources/logo.jpg', 'rb') as file_binary:
    text_binary = file_binary.readlines()

my_genrator = check_if_the_msg_in_line()
for line in text_binary:
    line = str(line)
    check_if_the_msg_in_line(line)


#python isawesome welldone goodjob

TypeError: check_if_the_msg_in_line() takes 0 positional arguments but 1 was given